File: android/app/src/main/java/com/faceauthoffline/FaceAuthModule.kt
package com.faceauthoffline

import android.graphics.Bitmap
import android.util.Base64
import android.util.Log
import com.facebook.react.bridge.*
import kotlinx.coroutines.*

class FaceAuthModule(reactContext: ReactApplicationContext) : ReactContextBaseJavaModule(reactContext) {
  private lateinit var tfliteEngine: TFLiteInferenceEngine
  private lateinit var livenessDetector: LivenessDetector
  private lateinit var faceEmbeddingEngine: FaceEmbeddingEngine
  private val scope = CoroutineScope(Dispatchers.Default + Job())

  init {
    try {
      tfliteEngine = TFLiteInferenceEngine(reactContext)
      livenessDetector = LivenessDetector()
      faceEmbeddingEngine = FaceEmbeddingEngine(tfliteEngine)
      Log.d(TAG, "FaceAuthModule initialized successfully")
    } catch (e: Exception) {
      Log.e(TAG, "Failed to initialize FaceAuthModule: ${e.message}")
    }
  }

  override fun getName(): String = "FaceAuthModule"

  @ReactMethod
  fun initialize(promise: Promise) {
    try {
      val success = tfliteEngine.loadModels()
      promise.resolve(success)
    } catch (e: Exception) {
      promise.reject("INIT_ERROR", e.message)
    }
  }

  @ReactMethod
  fun detectFace(base64Image: String, promise: Promise) {
    scope.launch {
      try {
        val bitmap = decodeBase64ToBitmap(base64Image)
        val detected = tfliteEngine.detectFace(bitmap)
        val result = WritableNativeMap().apply {
          putBoolean("detected", detected.isNotEmpty())
          if (detected.isNotEmpty()) {
            val box = detected[0]
            val boundingBox = WritableNativeMap()
            boundingBox.putDouble("x", box.left.toDouble())
            boundingBox.putDouble("y", box.top.toDouble())
            boundingBox.putDouble("width", ((box.right - box.left) as Float).toDouble())
            boundingBox.putDouble("height", ((box.bottom - box.top) as Float).toDouble())
            putMap("boundingBox", boundingBox)
          }
        }
        promise.resolve(result)
      } catch (e: Exception) {
        promise.reject("DETECT_ERROR", e.message)
      }
    }
  }

  @ReactMethod
  fun computeEmbedding(base64Image: String, promise: Promise) {
    scope.launch {
      try {
        val bitmap = decodeBase64ToBitmap(base64Image)
        val embedding = faceEmbeddingEngine.computeEmbedding(bitmap)
        val result = WritableNativeMap().apply {
          if (embedding != null) {
            val array = WritableNativeArray()
            embedding.forEach { array.pushDouble(it.toDouble()) }
            putArray("embedding", array)
          } else {
            putString("error", "Failed to compute embedding")
          }
        }
        promise.resolve(result)
      } catch (e: Exception) {
        promise.reject("EMBEDDING_ERROR", e.message)
      }
    }
  }

  @ReactMethod
  fun computeLandmarks(base64Image: String, promise: Promise) {
    scope.launch {
      try {
        val bitmap = decodeBase64ToBitmap(base64Image)
        val landmarks = tfliteEngine.computeLandmarks(bitmap)
        val landmarksMap = WritableNativeMap()

        if (landmarks.leftEye is List<*>) {
          landmarksMap.putArray("leftEye", listToWritableArray(landmarks.leftEye as List<Any>))
        }
        if (landmarks.rightEye is List<*>) {
          landmarksMap.putArray("rightEye", listToWritableArray(landmarks.rightEye as List<Any>))
        }
        if (landmarks.mouth is List<*>) {
          landmarksMap.putArray("mouth", listToWritableArray(landmarks.mouth as List<Any>))
        }
        if (landmarks.nose is List<*>) {
          landmarksMap.putArray("nose", listToWritableArray(landmarks.nose as List<Any>))
        }

        val headPoseMap = WritableNativeMap()
        headPoseMap.putDouble("yaw", (landmarks.headPose as? Map<String, Any>)?.get("yaw")?.toString()?.toDoubleOrNull() ?: 0.0)
        headPoseMap.putDouble("pitch", (landmarks.headPose as? Map<String, Any>)?.get("pitch")?.toString()?.toDoubleOrNull() ?: 0.0)
        headPoseMap.putDouble("roll", (landmarks.headPose as? Map<String, Any>)?.get("roll")?.toString()?.toDoubleOrNull() ?: 0.0)
        landmarksMap.putMap("headPose", headPoseMap)

        val result = WritableNativeMap().apply {
          putMap("landmarks", landmarksMap)
        }
        promise.resolve(result)
      } catch (e: Exception) {
        promise.reject("LANDMARKS_ERROR", e.message)
      }
    }
  }

  @ReactMethod
  fun matchEmbedding(
    embedding: ReadableArray,
    storedEmbeddings: ReadableArray,
    promise: Promise
  ) {
    try {
      val queryEmbedding = readableArrayToFloatArray(embedding)
      var bestMatch: String? = null
      var bestConfidence = 0f

      for (i in 0 until storedEmbeddings.size()) {
        val stored = storedEmbeddings.getMap(i)
        val storedVector = readableArrayToFloatArray(stored.getArray("vector")!!)
        val similarity = cosineSimilarity(queryEmbedding, storedVector)
        if (similarity > bestConfidence) {
          bestConfidence = similarity
          bestMatch = stored.getString("userId")
        }
      }

      val result = WritableNativeMap().apply {
        putBoolean("matched", bestConfidence >= EMBEDDING_THRESHOLD)
        if (bestMatch != null) {
          putString("userId", bestMatch)
        }
        putDouble("confidence", bestConfidence.toDouble())
      }
      promise.resolve(result)
    } catch (e: Exception) {
      promise.reject("MATCH_ERROR", e.message)
    }
  }

  private fun decodeBase64ToBitmap(base64: String): Bitmap {
    val decodedBytes = Base64.decode(base64, Base64.DEFAULT)
    return android.graphics.BitmapFactory.decodeByteArray(decodedBytes, 0, decodedBytes.size)
  }

  private fun readableArrayToFloatArray(array: ReadableArray): FloatArray {
    val result = FloatArray(array.size())
    for (i in 0 until array.size()) {
      result[i] = array.getDouble(i).toFloat()
    }
    return result
  }

  private fun listToWritableArray(list: List<Any>): WritableArray {
    val array = WritableNativeArray()
    list.forEach { item ->
      when (item) {
        is Number -> array.pushDouble(item.toDouble())
        is String -> array.pushString(item)
        is Boolean -> array.pushBoolean(item)
        is Map<*, *> -> {
          val map = WritableNativeMap()
          (item as Map<String, Any>).forEach { (key, value) ->
            when (value) {
              is Number -> map.putDouble(key, value.toDouble())
              is String -> map.putString(key, value)
              is Boolean -> map.putBoolean(key, value)
              else -> map.putNull(key)
            }
          }
          array.pushMap(map)
        }
        else -> array.pushNull()
      }
    }
    return array
  }

  private fun arrayToWritable(points: List<Pair<Float, Float>>): WritableArray {
    val array = WritableNativeArray()
    points.forEach {
      array.pushMap(mapToWritable(mapOf("x" to it.first, "y" to it.second)))
    }
    return array
  }

  private fun mapToWritable(map: Map<String, Any>): WritableMap {
    return WritableNativeMap().apply {
      map.forEach { (key, value) ->
        when (value) {
          is String -> putString(key, value)
          is Number -> putDouble(key, value.toDouble())
          is Boolean -> putBoolean(key, value)
        }
      }
    }
  }

  private fun cosineSimilarity(a: FloatArray, b: FloatArray): Float {
    if (a.size != b.size) return 0f
    var dot = 0f
    var normA = 0f
    var normB = 0f
    for (i in a.indices) {
      dot += a[i] * b[i]
      normA += a[i] * a[i]
      normB += b[i] * b[i]
    }
    return dot / (Math.sqrt((normA * normB).toDouble())).toFloat()
  }

  override fun onCatalystInstanceDestroy() {
    super.onCatalystInstanceDestroy()
    scope.cancel()
  }

  companion object {
    private const val TAG = "FaceAuthModule"
    private const val EMBEDDING_THRESHOLD = 0.6f
  }
}

File: ios/FaceAuthModule.swift
import Foundation
import React
import Vision
import CoreML
import AVFoundation

@objc(FaceAuthModule)
class FaceAuthModule: NSObject {
  private var tfliteEngine: TFLiteInferenceEngine?
  private var livenessDetector: LivenessDetector?
  private var faceEmbeddingEngine: FaceEmbeddingEngine?

  override init() {
    super.init()
    do {
      tfliteEngine = try TFLiteInferenceEngine()
      livenessDetector = LivenessDetector()
      if let engine = tfliteEngine {
        faceEmbeddingEngine = FaceEmbeddingEngine(engine)
      }
      print("FaceAuthModule initialized successfully")
    } catch {
      print("Failed to initialize FaceAuthModule: \(error.localizedDescription)")
    }
  }

  @objc
  func initialize(_ resolve: @escaping RCTPromiseResolveBlock, rejecter reject: @escaping RCTPromiseRejectBlock) {
    do {
      let success = try tfliteEngine?.loadModels() ?? false
      resolve(success)
    } catch {
      reject("INIT_ERROR", error.localizedDescription, error)
    }
  }

  @objc
  func detectFace(_ base64Image: String, resolver resolve: @escaping RCTPromiseResolveBlock, rejecter reject: @escaping RCTPromiseRejectBlock) {
    DispatchQueue.global(qos: .default).async {
      do {
        guard let bitmap = self.decodeBase64ToImage(base64Image) else {
          reject("DECODE_ERROR", "Failed to decode image", nil)
          return
        }
        let detections = try self.tfliteEngine?.detectFace(bitmap) ?? []
        var result: [String: Any] = ["detected": !detections.isEmpty]
        if let detection = detections.first {
          result["boundingBox"] = [
            "x": Int(detection.origin.x),
            "y": Int(detection.origin.y),
            "width": Int(detection.size.width),
            "height": Int(detection.size.height)
          ]
        }
        resolve(result)
      } catch {
        reject("DETECT_ERROR", error.localizedDescription, error)
      }
    }
  }

  @objc
  func computeEmbedding(_ base64Image: String, resolver resolve: @escaping RCTPromiseResolveBlock, rejecter reject: @escaping RCTPromiseRejectBlock) {
    DispatchQueue.global(qos: .default).async {
      do {
        guard let bitmap = self.decodeBase64ToImage(base64Image) else {
          reject("DECODE_ERROR", "Failed to decode image", nil)
          return
        }
        if let embedding = self.faceEmbeddingEngine?.computeEmbedding(bitmap) {
          resolve(["embedding": embedding])
        } else {
          resolve(["error": "Failed to compute embedding"])
        }
      } catch {
        reject("EMBEDDING_ERROR", error.localizedDescription, error)
      }
    }
  }

  @objc
  func computeLandmarks(_ base64Image: String, resolver resolve: @escaping RCTPromiseResolveBlock, rejecter reject: @escaping RCTPromiseRejectBlock) {
    DispatchQueue.global(qos: .default).async {
      do {
        guard let bitmap = self.decodeBase64ToImage(base64Image) else {
          reject("DECODE_ERROR", "Failed to decode image", nil)
          return
        }
        let landmarks = try self.tfliteEngine?.computeLandmarks(bitmap)
        let result: [String: Any] = [
          "landmarks": [
            "leftEye": landmarks?.leftEye ?? [],
            "rightEye": landmarks?.rightEye ?? [],
            "mouth": landmarks?.mouth ?? [],
            "nose": ["x": landmarks?.nose.x ?? 0, "y": landmarks?.nose.y ?? 0],
            "headPose": [
              "yaw": landmarks?.headPose.yaw ?? 0,
              "pitch": landmarks?.headPose.pitch ?? 0,
              "roll": landmarks?.headPose.roll ?? 0
            ]
          ]
        ]
        resolve(result)
      } catch {
        reject("LANDMARKS_ERROR", error.localizedDescription, error)
      }
    }
  }

  @objc
  func matchEmbedding(_ embedding: [NSNumber], storedEmbeddings: [[String: Any]], resolver resolve: @escaping RCTPromiseResolveBlock, rejecter reject: @escaping RCTPromiseRejectBlock) {
    do {
      let queryEmbedding = embedding.map { $0.floatValue }
      var bestMatch: String?
      var bestConfidence: Float = 0
      let EMBEDDING_THRESHOLD: Float = 0.6

      for storedData in storedEmbeddings {
        guard let vectorArray = storedData["vector"] as? [NSNumber],
              let userId = storedData["userId"] as? String else {
          continue
        }
        let storedVector = vectorArray.map { $0.floatValue }
        let similarity = cosineSimilarity(queryEmbedding, storedVector)
        if similarity > bestConfidence {
          bestConfidence = similarity
          bestMatch = userId
        }
      }

      var result: [String: Any] = [
        "matched": bestConfidence >= EMBEDDING_THRESHOLD,
        "confidence": Double(bestConfidence)
      ]
      if let match = bestMatch {
        result["userId"] = match
      }
      resolve(result)
    } catch {
      reject("MATCH_ERROR", error.localizedDescription, error)
    }
  }

  private func decodeBase64ToImage(_ base64String: String) -> UIImage? {
    guard let data = Data(base64Encoded: base64String) else { return nil }
    return UIImage(data: data)
  }

  private func cosineSimilarity(_ a: [Float], _ b: [Float]) -> Float {
    guard a.count == b.count else { return 0 }
    var dot: Float = 0
    var normA: Float = 0
    var normB: Float = 0
    for i in 0..<a.count {
      dot += a[i] * b[i]
      normA += a[i] * a[i]
      normB += b[i] * b[i]
    }
    return dot / (sqrt(normA * normB) + Float.leastNormalMagnitude)
  }

  @objc
  static func requiresMainQueueSetup() -> Bool {
    return false
  }
}
